src/compiler/transformers/esDecorators.ts

```diff
@@ -0,0 +1,2082 @@
+            };
+        }
+
+        function containsLexicalSuperInStaticInitializer(node: ClassLikeDeclaration) {
+            for (const member of node.members) {
+                if (isClassStaticBlockDeclaration(member) ||
+                    isPropertyDeclaration(member) && hasStaticModifier(member)) {
+                    if (member.transformFlags & TransformFlags.ContainsLexicalSuper) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        function transformClassLike(node: ClassLikeDeclaration, className: Expression) {
+            startLexicalEnvironment();
+
+            const classReference = node.name ?? factory.getGeneratedNameForNode(node);
+            const classInfo = createClassInfo(node);
+            const classDefinitionStatements: Statement[] = [];
+            let leadingBlockStatements: Statement[] | undefined;
+            let trailingBlockStatements: Statement[] | undefined;
+            let syntheticConstructor: ConstructorDeclaration | undefined;
+            let heritageClauses: NodeArray<HeritageClause> | undefined;
+            let shouldTransformPrivateStaticElementsInClass = false;
+
+            // 1. Class decorators are evaluated outside of the private name scope of the class.
+            const classDecorators = transformAllDecoratorsOfDeclaration(getAllDecoratorsOfClass(node));
+            if (classDecorators) {
+                // - Since class decorators don't have privileged access to private names defined inside the class,
+                //   they must be evaluated outside of the class body.
+                // - Since a class decorator can replace the class constructor, we must define a variable to keep track
+                //   of the mutated class.
+                // - Since a class decorator can add extra initializers, we must define a variable to keep track of
+                //   extra initializers.
+                classInfo.classDecoratorsName = factory.createUniqueName("_classDecorators", GeneratedIdentifierFlags.Optimistic);
+                classInfo.classDescriptorName = factory.createUniqueName("_classDescriptor", GeneratedIdentifierFlags.Optimistic);
+                classInfo.classExtraInitializersName = factory.createUniqueName("_classExtraInitializers", GeneratedIdentifierFlags.Optimistic);
+                classInfo.classThis = factory.createUniqueName("_classThis", GeneratedIdentifierFlags.Optimistic);
+                classDefinitionStatements.push(
+                    createLet(classInfo.classDecoratorsName, factory.createArrayLiteralExpression(classDecorators)),
+                    createLet(classInfo.classDescriptorName),
+                    createLet(classInfo.classExtraInitializersName, factory.createArrayLiteralExpression()),
+                    createLet(classInfo.classThis),
+                );
+
+                for (const member of node.members) {

```

Comment:
why is this loop here instead of merged into the one in createClassInfo?

Reply:
Mostly because `createClassInfo` doesn't have side-effects, but this loop does. I can move part of it into `createClassInfo` though, and have the side-effect remain in `transformClassLike` without needing the loop.